reconcile/external_resources/manager.py

```diff
@@ -0,0 +1,423 @@
+                    key=state.key,
+                    image=image,
+                    input=input,
+                )
+                self.state_mgr.set_external_resource_state(key, state)
+
+    def handle_deleted_resources(
+        self,
+        specs: Iterable[ExternalResourceSpec],
+    ) -> None:
+        pass
+        state_resource_keys = self.state_mgr.get_all_resource_keys()
+        desired_resource_keys = [ExternalResourceKey.from_spec(s) for s in specs]
+        diff = set(state_resource_keys) - set(desired_resource_keys)
+        for key in diff:
+            state = self.state_mgr.get_external_resource_state(key)
+            if state.status == ResourceStatus.DELETE_IN_PROGRESS:
+                reconcile_status = self.impl.get_resource_reconcile_status(key)
+                if reconcile_status != ReconcileStatus.IN_PROGRESS:
+                    state.ts = datetime.now(timezone.utc)
+                    if reconcile_status == ReconcileStatus.SUCCESS:
+                        state.status = ResourceStatus.DELETED
+                    elif reconcile_status == ReconcileStatus.ERROR:
+                        state.status = ResourceStatus.ERROR
+                    logging.info("Updating resource status to " + state.status)
+
+            # read how the resource was reconciled to get an input
+            if self.resource_can_be_deleted(state):
+                self.state_mgr.read_resource_reconcile_data(key)
+                if state.reconcile_data:
+                    image = state.reconcile_data.image
+                    input = state.reconcile_data.input
+                    self.impl.reconcile_resource(
+                        key=key,
+                        image=image,
+                        input=input,
+                        action=Action.DESTROY,
+                        dry_run=self.dry_run,
+                    )
+                    state.ts = datetime.now(timezone.utc)
+                    state.status = ResourceStatus.DELETE_IN_PROGRESS
+            elif state.status == ResourceStatus.DELETED:
+                logging.info("Removing DELETED resource from the index")
+                self.state_mgr.del_external_resource_state(key)
+
+    def get_provider_image(self, ers: ExternalResourceSpec) -> str:
+        key = f"{ers.provision_provider}-{ers.provider}"
+        module = self.modules.get(key)
+        if not module:
+            raise ValueError(key)
+        # TODO:Logic to override the image:version
+        return f"{module.image}:{module.default_version}"
+
+    def build_external_resource(self, ers: ExternalResourceSpec) -> ExternalResource:
+        resource = self.providers.get_factory(
+            ers.provision_provider
+        ).create_external_resource(ers)
+        return resource
+
+    def serialize_resource_input(self, resource: ExternalResource) -> str:
+        return base64.b64encode(json.dumps(resource.dict()).encode()).decode()
```

Comment:
we have one similar function in https://github.com/app-sre/qontract-reconcile/blob/8b4d20b528fa91849022d4726e1c5563afd99e9a/reconcile/utils/openshift_resource.py#L715, worth extract and reuse.